File: LinkedList-Stack-Queue/Q206.py
# https://leetcode.com/problems/reverse-linked-list/

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        # Iterative
        reversed = None
        while head:
            next = head.next
            head.next = reversed
            reversed = head
            head = next
        return reversed

    def reverseList(self, head: ListNode) -> ListNode:
        # Recursive
        if not head or not head.next:
            return head

        newTail = head.next
        newHead = self.reverseList(head.next)
        newTail.next = head
        head.next = None
        return newHead

File: LinkedList-Stack-Queue/Q206.js
/**
 * https://leetcode.com/problems/reverse-linked-list/
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head) {
    // Recursive
    if (!head || !head.next) {
        return head;
    }
    const newTail = head.next;
    const newHead = reverseList(head.next);
    newTail.next = head;
    head.next = null;
    return newHead;
};

var reverseList = function (head) {
    // Iterative
    let reversed = null, temp;
    while (head) {
        temp = head.next;
        head.next = reversed;
        reversed = head;
        head = temp;
    }
    return reversed;
};
